packages/jest-circus/src/utils.js

```diff
@@ -0,0 +1,225 @@
+/**
+ * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @flow
+ */
+
+import type {
+  AsyncFn,
+  BlockMode,
+  BlockName,
+  DescribeBlock,
+  Exception,
+  Hook,
+  SharedHookType,
+  Test,
+  TestContext,
+  TestFn,
+  TestMode,
+  TestName,
+  TestResults,
+} from '../types';
+
+const makeDescribe = (
+  name: BlockName,
+  parent: ?DescribeBlock,
+  mode?: BlockMode,
+): DescribeBlock => {
+  if (parent && !mode) {
+    // If not set explicitly, inherit from the parent describe.
+    mode = parent.mode;
+  }
+
+  return {
+    children: [],
+    hooks: [],
+    mode,
+    name,
+    parent,
+    tests: [],
+  };
+};
+
+const makeTest = (
+  fn: ?TestFn,
+  mode: TestMode,
+  name: TestName,
+  parent: DescribeBlock,
+): Test => {
+  if (!fn) {
+    mode = 'skip'; // skip test if no fn passed
+  } else if (!mode) {
+    // if not set explicitly, inherit from its parent describe
+    mode = parent.mode;
+  }
+
+  return {
+    duration: null,
+    errors: [],
+    fn,
+    mode,
+    name,
+    parent,
+    startedAt: null,
+    status: null,
+  };
+};
+
+const getAllHooksForDescribe = (
+  describe: DescribeBlock,
+): {[key: 'beforeAll' | 'afterAll']: Array<Hook>} => {
+  const result = {afterAll: [], beforeAll: []};
+
+  for (const hook of describe.hooks) {
+    switch (hook.type) {
+      case 'beforeAll':
+        result.beforeAll.push(hook);
+        break;
+      case 'afterAll':
+        result.afterAll.push(hook);
+        break;
+    }
+  }
+
+  return result;
+};
+
+const getEachHooksForTest = (
+  test: Test,
+): {[key: 'beforeEach' | 'afterEach']: Array<Hook>} => {
+  const result = {afterEach: [], beforeEach: []};
+  let {parent: block} = test;
+
+  do {
+    for (const hook of block.hooks) {
+      switch (hook.type) {
+        case 'beforeEach':
+          // Before hooks are executed from top to bottom, the opposite of the
+          // way we traversed it.
+          result.beforeEach.unshift(hook);
+          break;
+        case 'afterEach':
+          result.afterEach.push(hook);
+          break;
+      }
+    }
+  } while ((block = block.parent));
```

Comment:
yeah, i had a few ideas, but though they were premature optimization. i don't think we should worry about it yet